Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,592 @@
+//******************************************************************************
+//
+// Copyright (c) Microsoft. All rights reserved.
+//
+// This code is licensed under the MIT License (MIT).
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import <Foundation/NSDecimal.h>
+#import <StubReturn.h>
+#include <algorithm>
+
+#define EXPONENT_MAX 127
+#define EXPONENT_MIN -128
+
+#define MANTISSA(decimal) decimal->_mantissa
+
+#define LENGTH(decimal) decimal->_length
+
+#define EXPONENT(decimal) decimal->_exponent
+
+#define NEGATIVE_FLAG(decimal) decimal->_isNegative
+
+#define SET_TO_ZERO(decimal) \
+    LENGTH(decimal) = 0;     \
+    EXPONENT(decimal) = 0;   \
+    NEGATIVE_FLAG(decimal) = 0
+
+#define IS_NEGATIVE(decimal) decimal->_isNegative
+
+#define IS_POSITIVE(decimal) !IS_NEGATIVE(decimal)
+
+#define IS_ZERO(decimal) (decimal->_length == 0) && IS_POSITIVE(decimal)
+
+#define INVALIDATE_DECIMAL(decimal) decimal->_reserved = 1
+
+#define VALIDATE_ARITHMETIC_INPUT(result, leftOperand, rightOperand)                \
+    if (NSDecimalIsNotANumber(leftOperand) || NSDecimalIsNotANumber(leftOperand)) { \
+        if (result) {                                                               \
+            INVALIDATE_DECIMAL(result);                                             \
+        }                                                                           \
+        return NSCalculationNoError;                                                \
+    }
+
+#define CHECK_DIVIDE_BY0(decimal)                           \
+    NSDecimal zero = { 0 };                                 \
```

Comment:
>NSDecimal zero [](start = 4, length = 14)

This is creating a variable called "zero" in the scope of wherever this macro is used. That means, for instance, that you can't use the macro twice in the same function.